### PR TITLE
pkg/node: delete tunnel routes when remote nodes are removed

### PR DESCRIPTION
**Pull request by ianvernon**
Fixes: #5421

Signed-off by: Ian Vernon <ian@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5752)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

File: pkg/node/manager.go [hunk 1]
```diff
@@ -300,6 +300,7 @@ func DeleteNode(ni Identity, routesTypes RouteType) {
 
 			deleteTunnelMapping(n.IPv4AllocCIDR)
 			deleteTunnelMapping(n.IPv6AllocCIDR)
+			deleteIPRoute(n)
```
Comment (raybejjani) [inline pkg/node/manager.go]: Should the routes be removed before the tunnel? I'm not sure which which leads to which.

**Review by raybejjani — APPROVED** [review]

**Review by raybejjani — DISMISSED** [review]
on chatting with @tgraf, apparently this deleteIPRoute won't do what we intend. In particular (and he'll have to explain this better), `deleteIPRoute` is for direct routing and we'd need to remove the route for the node's CIDR block (I think, I probably understood it incorrectly).

File: pkg/node/manager.go (hunk 1 above)
Comment (tgraf) [inline pkg/node/manager.go]: deleteIPRoute is currently tailored to direct routing rules. We have to reverse `replaceNodeRoute()`

**Review by tgraf — CHANGES_REQUESTED** [review]

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
From what I could grok from looking at the code, and making a bunch of unnecessary changes, I believe we just need to call `deleteIPRoute` with the node's IPv4 and IPv6 CIDRs. I'm unsure how `deleteIPRoute` relates only to direct routing? I did the following locally with the patch in this PR:

```
vagrant@k8s1:~/go/src/github.com/cilium/cilium$ ip route show | grep cilium_host
10.0.0.0/16 via 10.0.0.1 dev cilium_host src 10.0.0.1 
10.0.0.1 dev cilium_host scope link 
10.1.0.0/16 via 10.0.0.1 dev cilium_host src 10.0.0.1 
vagrant@k8s1:~/go/src/github.com/cilium/cilium$ kubectl get nodes 
NAME      STATUS    ROLES     AGE       VERSION
k8s1      Ready     <none>    2h        v1.11.3
k8s2      Ready     <none>    1h        v1.11.3
vagrant@k8s1:~/go/src/github.com/cilium/cilium$ kubectl delete node k8s2
node "k8s2" deleted
vagrant@k8s1:~/go/src/github.com/cilium/cilium$ kubectl get nodes 
NAME      STATUS    ROLES     AGE       VERSION
k8s1      Ready     <none>    2h        v1.11.3
vagrant@k8s1:~/go/src/github.com/cilium/cilium$ ip route show | grep cilium_host
10.0.0.0/16 via 10.0.0.1 dev cilium_host src 10.0.0.1 
10.0.0.1 dev cilium_host scope link 
```

Is this what is expected?

**Comment by tgraf** [comment]
> Is this what is expected?

Yes, this is what is being expected. So it seems to work by accident. The function was definitely not intended to delete arbitrary node routes. Looking at the exact implementation, it does in fact work reliably.

**Review by tgraf — APPROVED** [review]

**Pull request merged**